File: benchmarks/micro/StackScalarLoadForwarding.c
#include <stdint.h>
#include <unistd.h>

static int64_t accumulate(int64_t seed, int64_t salt) {
    int64_t local = seed;
    local = local + salt;
    int64_t first = local;
    local = first ^ (first >> 7);
    return local + first;
}

int main(void) {
    int64_t total = (int64_t)getpid();

    for (int32_t i = 0; i < 200000; i += 1) {
        total += accumulate(total % 1000000007, (int64_t)i);
        total %= 1000000007;
    }

    return total <= 0 ? 1 : 0;
}

File: benchmarks/micro/StackScalarLoadForwarding.rs
fn accumulate(seed: i64, salt: i64) -> i64 {
    let mut local = seed;
    local = local + salt;
    let first = local;
    local = first ^ (first >> 7);
    local + first
}

fn main() {
    let mut total = std::process::id() as i64;

    for i in 0_i32..200_000 {
        total += accumulate(total % 1_000_000_007, i as i64);
        total %= 1_000_000_007;
    }

    if total <= 0 {
        std::process::exit(1);
    }
}
